okapi-core/src/main/java/org/folio/okapi/util/TokenCache.java

```diff
@@ -0,0 +1,174 @@
+package org.folio.okapi.util;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.apache.logging.log4j.Logger;
+import org.folio.okapi.common.OkapiLogger;
+
+public class TokenCache {
+
+  public static final long DEFAULT_TTL = 3 * 60 * 1000L;
+  public static final int DEFAULT_MAX_SIZE = 10000;
+
+  private final Map<String, CacheEntry> cache;
+  private final long ttl;
+
+  private static final Logger logger = OkapiLogger.get(TokenCache.class);
+
+  /**
+   * Constructor using the provided TTL and maxSize.
+   * 
+   * @param ttl cache entry time to live in milliseconds
+   * @param maxSize the maximum number of entries that may be cached at once
+   */
+  private TokenCache(long ttl, int maxSize) {
+    logger.info("Initializing token cache w/ ttl: {}, maxSize: {}", ttl, maxSize);
+    this.ttl = ttl;
+
+    this.cache = Collections.synchronizedMap(new LruCache<>(maxSize));
+  }
+
+  /**
+   * Cache an entry.
+   * 
+   * @param method HTTP method
+   * @param path path pattern
+   * @param userId X-Okapi-User-Id header to cache
+   * @param xokapiPerms X-Okapi-Permissions header to cache
+   * @param keyToken the token to be used in the cache key, from the request which triggered the
+   *        call to mod-authtoken
+   * @param token access token to cache - from the mod-authtoken response
+   */
+  public void put(String tenant, String method, String path, String userId, String xokapiPerms,
+      String keyToken, String token) {
+    long now = System.currentTimeMillis();
+    CacheEntry entry = new CacheEntry(token, userId, xokapiPerms, now + ttl);
+    String key = genKey(method, path, userId, keyToken);
+    MetricsHelper.recordTokenCacheCached(tenant, method, path, userId);
+    logger.debug("Caching: {} -> {}", key, token);
+    cache.put(key, entry);
```

Comment:
The cache will grow to its max size and will never shrink because an expired token is only evicted if there is a get request. If there is a get request for an expired token then mod-authtoken is called and a new cache entry is created.

Consider to extend the put method to check the two oldest cache entries whether they are expired and to remove them.

Reply:
I'm not seeing any good way to get the two oldest cache entries w/o iterating the entire keySet...  Any ideas?

Reply:
In LruCache replace K by String and V by CacheEntry. Then do this:

    @Override
    public CacheEntry put(String key, CacheEntry value) {
      CacheEntry old = super.put(key, value);

      if (size() < 3) {
        return old;
      }

      Iterator<Map.Entry<String, CacheEntry>> it = entrySet().iterator();
      for (int i=0; i<2; i++) {
        String k = it.next();
        if (expired(k)) {
          remove(k);
        } else {
          break;
        }
      }

      return old;
    }

Reply:
I had to look at the javadocs again, but I now see that the iteration order is least-recently accessed to most-recently accessed, exactly what we want.  Thanks @julianladisch.